### PR TITLE
Gutenboarding: Fix vertical types

### PR DESCRIPTION
**Pull request by sirreal**
Small change to update vertical types based on API results.

### CODE REVIEW

File: client/landing/gutenboarding/stores/onboard/types.ts [hunk 1]
```diff
@@ -28,23 +28,32 @@ export function isFilledFormValue< T >( value: FormValue< T > ): value is T {
...
-	synonyms: string[];
 }
 
+/**
+ * Known vertical
+ */
 export interface ApiVertical {
 	is_user_input_vertical: false;
-	vertical_id: '';
+
```
Comment (sirreal) [inline client/landing/gutenboarding/stores/onboard/types.ts]:
```suggestion
```

File: client/landing/gutenboarding/stores/onboard/types.ts [hunk 2]
```diff
@@ -28,23 +28,32 @@ export function isFilledFormValue< T >( value: FormValue< T > ): value is T {
...
 
+/**
+ * An ad-hoc vertical object based on user input
+ *
+ * A vertical like this is included with vertical suggestions if the query term does not match any known vertical.
+ */
 export interface UserVertical {
 	is_user_input_vertical: true;
-	vertical_id: string;
+
```
Comment (sirreal) [inline client/landing/gutenboarding/stores/onboard/types.ts]:
```suggestion
```

File: client/landing/gutenboarding/stores/onboard/types.ts [hunk 3]
```diff
@@ -28,23 +28,32 @@ export function isFilledFormValue< T >( value: FormValue< T > ): value is T {
 
 export type Vertical = ApiVertical | UserVertical;
 
+/**
+ * An ad-hoc vertical object based on user input
+ *
+ * A vertical like this is included with vertical suggestions if the query term does not match any known vertical.
+ */
```
Comment (ockham) [inline client/landing/gutenboarding/stores/onboard/types.ts]: Thanks for the comment, it's quite helpful.

`UserVertical` strikes me as a bit off. It feels a bit like we're trying to impose the `Vertical` type onto something that just isn't one? (see missing `vertical_id`). 

I don't quite get where we're needing it :thinking: I thought it was for storing a user-supplied string in `siteVertical` state, if it didn't match any of our predefined verticals, but we're using a different type for that.

Aren't all `state.verticals` always `ApiVertical`s?

Comment (sirreal) [reply client/landing/gutenboarding/stores/onboard/types.ts]: This is all descriptive of what's coming back from the API, none of which is validated or normalized, something we should consider adding.

> UserVertical strikes me as a bit off.

While I tend to agree, this intends to describe what we get in the API response.

> Aren't all state.verticals always ApiVerticals?

Not based on what I observed, I've interpreted the vertical results in the way this type describes. You get a list of "Normal" verticals, and if your search doesn't match one, you also have this "ad-hoc" vertical returned in the same list as the rest of the verticals. It's a heterogeneous list.

Try something like `wpcom/v2/verticals?search=cat` in the [API console](https://developer.wordpress.com/docs/api/console/):

`foobar`:

Screen Shot 2019-11-15 at 10 01 12


`Fast Food Restaurant`:

Screen Shot 2019-11-15 at 10 02 49

An imperfect match returns the user + known verticals.

🤔 🤔 🤔 On second thought 🤔 🤔 🤔 

We're querying all known verticals up front, and not hitting the API with any search terms. In that case, verticals should, indeed, be consistently of type "API Vertical", and we could even normalize to ensure no "User Vertical" is provided.

Thanks for mentioning this point!

**Comment by sirreal** [comment]
I'll narrow the type of this, @ockham what are your thoughts on validating incoming data with something like https://github.com/Automattic/wp-calypso/blob/master/client/lib/make-json-schema-parser/index.js?

**Comment by sirreal** [comment]
Added comments and improved types several places. Ready for another review.

File: client/landing/gutenboarding/stores/onboard/types.ts (hunk 3 above)
Comment (ockham) [reply client/landing/gutenboarding/stores/onboard/types.ts]: Thanks a lot for your reply, and those very instructive examples! I second you conclusion! :smile:

**Comment by ockham** [comment]
> I'll narrow the type of this, @ockham what are your thoughts on validating incoming data with something like https://github.com/Automattic/wp-calypso/blob/master/client/lib/make-json-schema-parser/index.js?

Yes! It'd be lovely if we could re-use type information from our TS interfaces, and didn't have to duplicate them through some JS syntax :thinking: Bit of a similar problem as with Gutenberg block attributes. Compile-time vs runtime and all.

**Pull request merged**